src/unix/legacy.cc

```diff
@@ -23,6 +23,372 @@
+  /* directory.                                                         */
+  if(cwdPathLoc == 0)
+  {
+     if(chdir(cwdbuf) == -1)
+     {
+       /* errno will be set by chdir() failure.                         */
+       _openrc = -1;
+     }
+  }
+  else
+  {
+    if(chdir(*cwdpath) == -1)
+    {
+      /* errno will be set by chdir() failure.                          */
+      _openrc = -1;
+    }
+  }
+  
+leave:
+  if(*cwdpath != NULL) /* if malloc'd storage, free it now.             */
+    free(*cwdpath);
+    
+  return _openrc;
+}
+
+int ibmi_openat2(int dirfd, const char * pathname, int flags, mode_t mode)
+{
+  int _openrc = -1;                         /* default to fail          */
+  const unsigned int cwdBufSize = 4096;     /* cwd init buffer size     */
+  char cwdbuf[cwdBufSize];                  /* cwd init buffer          */
+  char ** cwdpath = NULL;                   /* cwd malloc'd path        */
+  unsigned int cwdPathLoc = 0;              /* cwd path location        */
+  
+  if(pathname == (const char *)NULL)        /* safety/sanity check      */
+  {
+    errno = EINVAL;
+    return -1;
+  }
+  
+  /* path is absolute, open it.                                         */
+  if(*pathname == '/')
+    return open(pathname, flags, mode);
+  
+  /* path is relative, open in current working directory of process.    */
+  if(dirfd == AT_FDCWD)
+    return open(pathname, flags, mode);
+  
+  /* path is relative, but caller doesn't want to use process's current */
+  /* working directory. Build the current working directory, so it      */
+  /* can be changed back to.                                            */
+  _openrc = getcwdpath((const char *)&cwdbuf, cwdBufSize,
+                        cwdpath, &cwdPathLoc);
+  
+  if(_openrc == -1)
+  {
+    /* errno will be set by getcwdpath() failure.                       */
+    return -1;
+  }
+  
+  /* path is relative, change process's current working directory to    */
+  /* the directory file descriptor.                                     */
+  if(fchdir(dirfd) == -1)
+  {
+    _openrc = -1;
+    goto leave;
+  }
+  
+  /* path is relative, open it relative to directory file descriptor.   */
+  _openrc = open(pathname, flags, mode);
+  
+  /* change working directory back to original process's working        */
+  /* directory.                                                         */
+  if(cwdPathLoc == 0)
+  {
+     if(chdir(cwdbuf) == -1)
+     {
+       /* errno will be set by chdir() failure.                         */
+       _openrc = -1;
+     }
+  }
+  else
+  {
+    if(chdir(*cwdpath) == -1)
+    {
+      /* errno will be set by chdir() failure.                          */
+      _openrc = -1;
+    }
+  }
+  
+leave:
+  if(*cwdpath != NULL) /* if malloc'd storage, free it now.             */
+    free(*cwdpath);
+    
+  return _openrc;
+}
+
+#define fstatat ibmi_fstatat
+#define openat ibmi_openat
+#endif
```

Comment:
Could you move this stuff to a separate file and only include it in the build on aix/os400 systems?